### PR TITLE
Implement GPU kernel for cross entropy operator.

### PR DESCRIPTION
**Pull request by qingqing01**
#3492 

- Implement GPU kernel for cross entropy operator.
- Fix the origin CPU implementation.

### CODE REVIEW

File: paddle/operators/cross_entropy_op.cu [hunk 1]
```diff
@@ -12,10 +12,123 @@
...
+#include "paddle/framework/op_registry.h"
+#include "paddle/platform/assert.h"
+
+namespace paddle {
+namespace operators {
+
+using Tensor = framework::Tensor;
+
+template <typename T>
+struct clipping_log {
```
Comment (jacquesqiao) [inline paddle/operators/cross_entropy_op.cu]: why not just use a function
```cpp
template<typename T> 
__host__ __device__ T clipping_log(const Tx) {
  ...
  return xx;
}
```

File: paddle/operators/cross_entropy_op.cu [hunk 2]
```diff
@@ -12,10 +12,123 @@
...
+
+template <typename T>
+__global__ void CrossEntropyKernel(T* Y, const T* X, const int* label,
+                                   const int N, const int D) {
+  // TOOD(qingqing) define CUDA_1D_KERNEL_LOOP macro in a common file.
+  // CUDA_1D_KERNEL_LOOP(i, N) {
+  for (int i = blockIdx.x * blockDim.x + threadIdx.x; i < N;
+       i += blockDim.x * gridDim.x) {
+    PADDLE_ASSERT(label[i] >= 0 && label[i] < D);
+    Y[i] = -clipping_log<T>()(X[i * D + label[i]]);
```
Comment (QiJune) [inline paddle/operators/cross_entropy_op.cu]: 我有点没搞清楚啊，交叉熵的计算公式是什么呀？不是 cost = 求和(yln(x) + (1-y)ln(1-x)) 这个吗？

File: paddle/operators/cross_entropy_op.cu (hunk 1 above)
Comment (qingqing01) [inline paddle/operators/cross_entropy_op.cu]: I notice that Majel and WarpCTC use many [`functors`](https://stackoverflow.com/questions/356950/c-functors-and-their-uses) to define many common functions. The functor has its own advantages, see:  https://stackoverflow.com/questions/356950/c-functors-and-their-uses

But it is not necessary here, I change it to the plain function.

File: paddle/operators/cross_entropy_op.cu (hunk 2 above)
Comment (qingqing01) [inline paddle/operators/cross_entropy_op.cu]: label实际是one-hot类型: [0, 0, ..., 1, ..., 0,0] ，所以这里是：y[i] =  -log(x[ i * D + label[i] ]), 求和通常不在`cross_entropy_op`里做，是后面接的op或者打印loss时再求和。

**Comment by zchen0211** [comment]
Hello, do we need something like <<<grid, block, 0, stream_id >>> to execute it?

**Comment by qingqing01** [comment]
@zchen0211 We need to specify the stream, like <<<grid, block, 0, stream_id >>>.  I think the stream_id can be got from `CUDAContext`, but there is no stream in `CUDAContext` now. And @QiJune said he has worked on it last week.
